File: leetcode/easy/DiameterOfBinaryTree.py
from typing import Optional


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        """
        Calculates the diameter of a binary tree.

        The diameter is the length of the longest path between any two nodes in a tree.
        This path may or may not pass through the root. The length is the number of edges.

        This solution uses a recursive Depth-First Search (DFS) approach. For each node,
        we calculate two things:
        1. The depth of the node: The longest path (number of edges) from this node
           downwards to a leaf. This value is returned to the parent caller.
        2. The diameter passing through the node: This is the sum of the depths of its
           left and right subtrees. This value is a candidate for the overall maximum
           diameter and is used to update a global tracker.
        """
        # Use an instance variable to store the maximum diameter found.
        self.diameter = 0

        def dfs(node: Optional[TreeNode]) -> int:
            """
            Recursively calculates the depth of a subtree.
            Returns: The depth of the subtree rooted at `node` (number of edges).
            Updates: `self.diameter` with the longest path found so far.
            """
            # The depth of a null tree is -1. This base case makes the calculation
            # for a leaf node (1 + max(-1, -1)) correctly result in 0.
            if not node:
                return -1

            # Recursively find the depth of the left and right subtrees.
            left_depth = dfs(node.left)
            right_depth = dfs(node.right)

            # The diameter at the current node is the sum of the depths of its left and right children, plus the two edges connecting them to this node.
            self.diameter = max(self.diameter, left_depth + right_depth + 2)

            # The depth of the current node is 1 + the depth of its deeper child.
            return 1 + max(left_depth, right_depth)

        dfs(root)
        return self.diameter

File: leetcode/easy/DiameterOfBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // This variable will store the maximum diameter found anywhere in the tree.
    // It's initialized to 0 because a single node has a diameter of 0.
    private int maxDiameter = 0;

    /**
     * Calculates the diameter of a binary tree.
     * The diameter of a binary tree is the length of the longest path between
     * any two nodes in the tree. This path may or may not pass through the root.
     *
     * This solution uses a single-pass Depth-First Search (DFS) approach.
     * For each node, we recursively calculate the height of its left and right
     * subtrees. While doing so, we also calculate the diameter passing through
     * the current node and update a global maximum diameter.
     *
     * Time Complexity: O(N), where N is the number of nodes in the tree, as each
     *                  node is visited exactly once.
     * Space Complexity: O(H), where H is the height of the tree, for the recursion
     *                   stack. In the worst case (a skewed tree), this can be O(N).
     *
     * @param root The root node of the binary tree.
     * @return The length of the longest path between any two nodes.
     */
    public int diameterOfBinaryTree(TreeNode root) {
        // Reset maxDiameter for each new call to ensure correctness if called multiple times.
        maxDiameter = 0;
        // Call the helper function which calculates height and updates maxDiameter.
        calculateHeight(root);
        return maxDiameter;
    }

    /**
     * A recursive helper function that calculates the height of a subtree
     * and simultaneously updates the maximum diameter found so far.
     *
     * This function performs a post-order traversal: it first processes children
     * to get their heights, then processes the current node.
     *
     * @param node The current node being processed.
     * @return The height of the subtree rooted at the current node.
     */
    private int calculateHeight(TreeNode node) {
        // Base case: An empty tree has a height of 0.
        if (node == null) {
            return 0;
        }

        // Recursively calculate the height of the left subtree.
        int leftHeight = calculateHeight(node.left);
        // Recursively calculate the height of the right subtree.
        int rightHeight = calculateHeight(node.right);

        // Calculate the diameter passing through the current node.
        // This is the sum of the heights of its left and right subtrees.
        // Update the global maximum diameter if this path is longer.
        maxDiameter = Math.max(maxDiameter, leftHeight + rightHeight);

        // Return the height of the current node's subtree.
        // The height of a node is 1 (for the node itself) + the maximum height
        // of its left or right subtree.
        return 1 + Math.max(leftHeight, rightHeight);
    }
}
